block_object.go

```diff
@@ -1,44 +1,132 @@
+
+type blockObject interface {
+	validateType() MessageObjectType
 }
 
-// validateType enforces block objects for element and block parameters
-func (s ImageBlockObject) validateType() MessageObjectType {
-	return s.Type
+type BlockObjects struct {
+	TextObjects         []*TextBlockObject
+	ConfirmationObjects []*ConfirmationBlockObject
+	OptionObjects       []*OptionBlockObject
+	OptionGroupObjects  []*OptionGroupBlockObject
+}
+
+// UnmarshalJSON implements the Unmarshaller interface for BlockObjects, so that any JSON
+// unmarshalling is delegated and proper type determination can be made before unmarshal
+func (b *BlockObjects) UnmarshalJSON(data []byte) error {
+	var raw []json.RawMessage
+	err := json.Unmarshal(data, &raw)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range raw {
+		var obj map[string]interface{}
+		err := json.Unmarshal(r, &obj)
+		if err != nil {
+			return err
+		}
+
+		blockObjectType := getBlockObjectType(obj)
+
+		switch blockObjectType {
+		case PlainTextType, MarkdownType:
+			object, err := unmarshalBlockObject(r, &TextBlockObject{})
+			if err != nil {
+				return err
+			}
+			b.TextObjects = append(b.TextObjects, object.(*TextBlockObject))
+		case motConfirmation:
+			object, err := unmarshalBlockObject(r, &ConfirmationBlockObject{})
+			if err != nil {
+				return err
+			}
+			b.ConfirmationObjects = append(b.ConfirmationObjects, object.(*ConfirmationBlockObject))
+		case motOption:
+			object, err := unmarshalBlockObject(r, &OptionBlockObject{})
+			if err != nil {
+				return err
+			}
+			b.OptionObjects = append(b.OptionObjects, object.(*OptionBlockObject))
+		case motOptionGroup:
+			object, err := unmarshalBlockObject(r, &OptionGroupBlockObject{})
+			if err != nil {
+				return err
+			}
+			b.OptionGroupObjects = append(b.OptionGroupObjects, object.(*OptionGroupBlockObject))
+
+		}
+	}
+
+	return nil
+}
+
+// Ideally would have a better way to identify the block objects for
+// type casting at time of unmarshalling, should be adapted if possible
+// to accomplish in a more reliable manner.
+func getBlockObjectType(obj map[string]interface{}) string {
+	if t, ok := obj["type"].(string); ok {
+		return t
+	}
+	if _, ok := obj["confirm"].(string); ok {
+		return "confirm"
+	}
+	if _, ok := obj["options"].(string); ok {
+		return "option_group"
+	}
+	if _, ok := obj["text"].(string); ok {
+		if _, ok := obj["value"].(string); ok {
+			return "option"
+		}
+	}
+	return ""
 }
 
-// NewImageBlockObject returns a new instance of an image block element
-func NewImageBlockObject(imageURL, altText string) *ImageBlockObject {
-	return &ImageBlockObject{
-		Type:     motImage,
-		ImageURL: imageURL,
-		AltText:  altText,
+func unmarshalBlockObject(r json.RawMessage, object blockObject) (blockObject, error) {
+	err := json.Unmarshal(r, object)
+	if err != nil {
+		return nil, err
+	}
+	return object, nil
+}
+
+func (o *BlockObjects) appendToBlockObjects(appendObjects []blockObject) {
```

Comment:
this function seems entirely unnecessary, also seems unused now.

Reply:
might indeed be unnecessary now, will confirm tomorrow 👍

Reply:
removed